python/pyspark/sql/types.py

```diff
@@ -57,7 +57,25 @@ def __ne__(self, other):
 
     @classmethod
     def typeName(cls):
-        return cls.__name__[:-4].lower()
+        typeTypeNameMap = {"DataType": "data",
+                           "NullType": "null",
+                           "StringType": "string",
+                           "BinaryType": "binary",
+                           "BooleanType": "boolean",
+                           "DateType": "date",
+                           "TimestampType": "timestamp",
+                           "DecimalType": "decimal",
+                           "DoubleType": "double",
+                           "FloatType": "float",
+                           "ByteType": "byte",
+                           "IntegerType": "integer",
+                           "LongType": "long",
+                           "ShortType": "short",
+                           "ArrayType": "array",
+                           "MapType": "map",
+                           "StructField": "struct",
```

Comment:
@viirya The way I found this bug: 
I wanted to figure out the schema of a dataset. I loaded it into a data frame  and asked its schema. Then I called the `typeName` on each column. I do not know this is/was best way to do this, but I think it is valid to call `typeName` against a `dataType` to get its real type.